types/coin.go

```diff
@@ -145,66 +154,154 @@ func (coins Coins) IsValid() bool {
 // {2A} + {A, 2B} = {3A, 2B}
 // {2A} + {0B} = {2A}
 //
+// NOTE: Plus makes no assumptions about the order or unique contents of each
```

Comment:
We should enforce that coinsets are sorted by denom. Then you get faster arithmetic

Reply:
@ValarDragon enforce....do you mean explicitly sort here or panic if they're not sorted? We might as well sort if we're just checking (worst case runtime).

Reply:
It should be an invariant of all coins that the denoms are sorted. So this statement:
>  // NOTE: Plus makes no assumptions about the order or unique contents of each 
would then become
>  // NOTE: Plus uses the fact that coins are sorted by denoms

Reply:
I'm agnostic as to whether this is enforced here or not. As I mentioned via DM's, I think the optimal thing to do would be to have "arithmetic coins" / "intermediate coins" which is the result of any arithmetic and can be negative. Then you have a method to cast to coins when you want to put in state. (And all things coming from txs would also be Coins, so you get the uint properties there)

This lets you do `A - B + C` nicely, amongst other things.

Reply:
But we're not enforcing this invariant. So you simply want it stated?

w.r.t impl, perhaps. That would require two types. I think having a simple map would make things easier and we would still be able to have the `uint` type. The `sum` method would be very simple.

Reply:
How do you do `a - b + c` with uints, allowing the result to be negative at `a - b`? We panic when it becomes negative.

I want this invariant to be enforced in the conversion layer from arithmetic coins to normal coins and vice-versa.

Reply:
> You don't. You check before the operation (using a map).

So it'd panic? That seems like it will end up being hard to catch / definitely cause problems in production. Sounds good for doing arithmetic coins in a separate PR.

Reply:
Yah, it'd panic as soon as it sees such a case. What are your opinions on the current (temporary) solution?

Reply:
Temporary solution seems better than what we already have from safety, but worse usability. It is fine for its own PR, since arithmetic coins is a genuinely separate thing. I do think arithmetic coins is necessitated by the switch to uints, so perhaps the order should be the reverse. (Start arithmetic coins, and then make uints) But then again, we have launch pressures, so I guess arithmetic coins can be postlaunch.

Or were you asking about the actual implementation of the PR, I haven't looked at any of the code changes except for this function.

Reply:
Actually were also being slower, so that kinda sucks. I do still want us to maintain denoms being sorted.

Reply:
Confused by your last statement?

Reply:
Coin denominations being sorted yields faster addition rules.

Reply:
I see. Let's continue your thoughts and discussion here: #1273 :-)